File: Python/LinkedList/Rotate a Linked List.py
//https://www.geeksforgeeks.org/problems/rotate-a-linked-list/1
class Solution:
    
    #Function to rotate a linked list.
    def rotate(self, head, k):
        # code here
        if not head or not head.next or k == 0:
            return head
        
        length = 1
        temp = head
        while temp.next:
            temp = temp.next
            length += 1
        
        k = k % length
        if k == 0:
            return head
            
        temp.next = head
        
        newTail = head
        for i in range(1, k):
            newTail = newTail.next
        
        newHead = newTail.next
        newTail.next = None # breaking the circular connection
        return newHead

File: Java/Linked List/Rotate a Linked List.java
//https://www.geeksforgeeks.org/problems/rotate-a-linked-list/1
class Solution {
    public Node rotate(Node head, int k) {
        // add code here
        if (head == null || head.next == null || k == 0)
            return head;

        // Step 1: Find the length of the linked list
        Node temp = head;
        int length = 1;
        while (temp.next != null) {
            temp = temp.next;
            length++;
        }

        // Step 2: Normalize k
        k = k % length;
        if (k == 0)
            return head;

        // Step 3: Make it circular
        temp.next = head;

        // Step 4: Traverse to the (k)th node
        Node newTail = head;
        for (int i = 1; i < k; i++) {
            newTail = newTail.next;
        }

        // Step 5: Set new head and break the circle
        Node newHead = newTail.next;
        newTail.next = null;

        return newHead;
    }
}
